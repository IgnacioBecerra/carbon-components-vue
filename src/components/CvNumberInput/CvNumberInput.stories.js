import { action } from '@storybook/addon-actions';
import {
  sbCompPrefix,
  storyParametersObject,
} from '../../global/storybook-utils';
import { StorybookGroupConst } from './StorybookGroupConst';
import CvNumberInput from './CvNumberInput.vue';

export default {
  title: `${sbCompPrefix}/${StorybookGroupConst}/CvNumberInput`,
  component: CvNumberInput,
  argTypes: {
    // attrs
    ariaLabelForDownButton: {
      type: 'string',
      table: {
        type: { summary: 'string' },
        category: 'attributes',
      },
      description: 'Aria label for down (decrease) button',
    },
    ariaLabelForUpButton: {
      type: 'string',
      table: {
        type: { summary: 'string' },
        category: 'attributes',
      },
      description: 'Aria label for up (increase) button',
    },
    disabled: {
      type: 'boolean',
      table: {
        type: { summary: 'boolean' },
        category: 'attributes',
      },
      description: 'Specify whether the `<input>` should be disabled',
    },
    max: {
      type: 'string',
      table: {
        type: { summary: 'string | number' },
        category: 'attributes',
      },
      description: '`max` attribute for input HTML element',
    },
    min: {
      type: 'string',
      table: {
        type: { summary: 'string | number' },
        category: 'attributes',
      },
      description: '`min` attribute for input HTML element',
    },
    step: {
      type: 'string',
      table: {
        type: { summary: 'string | number' },
        category: 'attributes',
      },
      description: '`step` attribute for input HTML element',
    },
    // props
    formItem: {
      type: 'boolean',
      table: {
        type: { summary: 'boolean' },
        category: 'props',
        defaultValue: true,
      },
      description:
        "Adds .bx--form-item class to component's wrapping HTML element",
    },
    helperText: {
      type: 'string',
      table: {
        type: { summary: 'string' },
        category: 'props',
      },
      description:
        'Provide text that is used alongside the control label for additional help',
    },
    invalidMessage: {
      type: 'string',
      table: {
        type: { summary: 'string' },
        category: 'props',
      },
      description:
        'Provide the text that is displayed when the control is in an invalid state',
    },
    label: {
      type: 'string',
      table: {
        type: { summary: 'string' },
        category: 'props',
      },
      description: "Input's label",
    },
    light: {
      type: 'boolean',
      table: {
        type: { summary: 'boolean' },
        category: 'props',
        defaultValue: false,
      },
      description:
        "Toggles light version. For use on `$ui-01` backgrounds only. Don't use this to make tile background color same as container background color.",
    },
    modelValue: {
      type: 'string',
      table: {
        type: { summary: 'string' },
        category: 'props',
      },
      description:
        "Input's value, modelValue is the vue3 default 'prop' for two-way data binding with v-model",
    },
    warnText: {
      type: 'string',
      table: {
        type: { summary: 'string' },
        category: 'props',
      },
      description:
        'Provide the text that is displayed when the control is in warning state',
    },
    // slots
    'label ': {
      type: 'string',
      table: {
        type: { summary: 'string | html | Component' },
        category: 'slots',
      },
      description: 'Input label slot',
    },
    'helper-text': {
      type: 'string',
      table: {
        type: { summary: 'string | html | Component' },
        category: 'slots',
      },
      description:
        "Helper text slot. It isn't shown if invalid message or warn text, either props or slots, are available",
    },
    'invalid-message': {
      type: 'string',
      table: {
        type: { summary: 'string | html | Component' },
        category: 'slots',
      },
      description: 'Invalid message slot.',
    },
    'warn-text': {
      type: 'string',
      table: {
        type: { summary: 'string | html | Component' },
        category: 'slots',
      },
      description:
        "Warn text slot. It isn't shown if invalid message, either prop or slot, is available",
    },
    // events
    'update:modelValue': {
      type: 'event',
      table: {
        type: { summary: 'event' },
        category: 'events',
      },
      control: { type: null },
      description:
        "Triggered on textarea update. `update:modelValue` is the vue3 default 'event' for two-way data binding with v-model",
    },
  },
};

const template = `<cv-number-input @change="onChange" @input="onInput" v-bind="args"></cv-number-input>`;
const Template = (args, { argTypes }) => {
  return {
    props: Object.keys(argTypes),
    components: { CvNumberInput },
    template,
    setup: () => ({
      args,
      onInput: action('input'),
      onChange: action('change'),
    }),
  };
};

export const Default = Template.bind({});
Default.parameters = storyParametersObject(
  Default.parameters,
  template,
  Default.args
);
