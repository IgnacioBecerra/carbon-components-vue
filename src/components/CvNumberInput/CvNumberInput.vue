<template>
  <cv-wrapper
    :tag-type="formItem ? 'div' : ''"
    class="cv-number-input"
    :class="`${carbonPrefix}--form-item`"
  >
    <div
      :class="[
        `${carbonPrefix}--number`,
        {
          [`${carbonPrefix}--number--light`]: isLight,
          [`${carbonPrefix}--number--helpertext`]: isHelper,
          [`cv-number-input`]: !formItem,
        },
      ]"
      :data-invalid="dataInvalidAttributeValue"
    >
      <label v-show="label" :for="cvId" :class="`${carbonPrefix}--label`">
        {{ label }}
      </label>

      <div
        :class="[
          `${carbonPrefix}--number__input-wrapper`,
          { [`${carbonPrefix}--number__input-wrapper--warning`]: isWarn },
        ]"
      >
        <input
          ref="input"
          v-model.number="internalValue"
          :id="cvId"
          :data-invalid="dataInvalidAttributeValue"
          type="number"
          v-bind="$attrs"
          :disabled="disabled"
          :step="step"
          :min="min"
          :max="max"
        />

        <WarningFilled16 v-if="isInvalid" :class="iconClasses" />

        <WarningAltFilled16 v-if="isWarn && !isInvalid" :class="iconClasses" />

        <div :class="`${carbonPrefix}--number__controls`">
          <button
            :class="`${carbonPrefix}--number__control-btn down-icon`"
            type="button"
            :aria-label="ariaLabelForDownButton"
            :disabled="disabled"
            aria-live="polite"
            aria-atomic="true"
            @click="doDown"
          >
            <Subtract16 class="down-icon" />
          </button>

          <div :class="`${carbonPrefix}--number__rule-divider`" />

          <button
            :class="`${carbonPrefix}--number__control-btn up-icon`"
            type="button"
            :aria-label="ariaLabelForUpButton"
            :disabled="disabled"
            aria-live="polite"
            aria-atomic="true"
            @click="doUp"
          >
            <Add16 class="up-icon" />
          </button>

          <div :class="`${carbonPrefix}--number__rule-divider`" />
        </div>
      </div>

      <div v-if="isInvalid" :class="`${carbonPrefix}--form-requirement`">
        <slot name="invalid-message">{{ invalidMessage }}</slot>
      </div>

      <div
        v-if="isWarn && !isInvalid"
        :class="`${carbonPrefix}--form-requirement`"
      >
        <slot name="warn-text">{{ warnText }}</slot>
      </div>

      <div
        v-if="!(isInvalid || isWarn) && isHelper"
        :class="`${carbonPrefix}--form__helper-text`"
      >
        <slot name="helper-text">{{ helperText }}</slot>
      </div>
    </div>
  </cv-wrapper>
</template>

<script setup>
import { computed, ref } from 'vue';
import Add16 from '@carbon/icons-vue/es/add/16';
import Subtract16 from '@carbon/icons-vue/es/subtract/16';
import WarningFilled16 from '@carbon/icons-vue/es/warning--filled/16';
import WarningAltFilled16 from '@carbon/icons-vue/es/warning--alt--filled/16';
import { useCvId, props as propsCvId } from '../../use/cvId';
import { props as propsTheme } from '../../use/cvTheme';
import { useCvInputHelpers } from '../../use/cvInput';
import { carbonPrefix } from '../../global/settings';
import CvWrapper from '../CvWrapper/CvWrapper';

const props = defineProps({
  disabled: Boolean,
  modelValue: {
    type: [String, Number],
    default: '',
  },
  formItem: {
    type: Boolean,
    default: true,
  },
  helperText: {
    type: String,
    default: undefined,
  },
  invalidMessage: {
    type: String,
    default: undefined,
  },
  label: {
    type: String,
    default: '',
  },
  ariaLabelForUpButton: {
    type: String,
    default: 'Increase number input',
  },
  ariaLabelForDownButton: {
    type: String,
    default: 'Decrease number input',
  },
  min: {
    type: [String, Number],
    default: undefined,
  },
  max: {
    type: [String, Number],
    default: undefined,
  },
  step: {
    type: [String, Number],
    default: undefined,
  },
  warnText: {
    type: String,
    default: undefined,
  },
  ...propsCvId,
  ...propsTheme,
});

const cvId = useCvId(props);
const { isInvalid, isWarn, isHelper, isLight } = useCvInputHelpers(props);
const dataInvalidAttributeValue = computed(() =>
  isInvalid ? undefined : true
);

const iconClasses = computed(() => {
  const classes = [];

  if (isInvalid || isWarn) {
    classes.push(`${carbonPrefix}--number__invalid`);
  }

  if (isWarn && !isInvalid) {
    classes.push(`${carbonPrefix}--number__invalid--warning`);
  }

  return classes.join(' ');
});

const emit = defineEmits(['update:modelValue']);
const input = ref(null);
const internalValue = computed({
  get() {
    const { modelValue } = props;
    if (typeof modelValue === 'number') return modelValue;
    const valueAsNumber = Number(modelValue);
    return Number.isNaN(valueAsNumber) ? '' : valueAsNumber;
  },
  set(value) {
    onInput(value);
  },
});

const onInput = value => {
  emit('update:modelValue', value);
};

const programmaticEventsToTrigger = ['input', 'change'];
const onProgrammaticValueChanged = inputRef => {
  programmaticEventsToTrigger.forEach(eventName => {
    const event = new Event(eventName);
    inputRef.value.dispatchEvent(event);
  });
};

const doUp = () => {
  input.value.stepUp();
  onProgrammaticValueChanged(input);
};

const doDown = () => {
  input.value.stepDown();
  onProgrammaticValueChanged(input);
};
</script>

<script>
export default {
  inheritAttrs: false,
};
</script>
