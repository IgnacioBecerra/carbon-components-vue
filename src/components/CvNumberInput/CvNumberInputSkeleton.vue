<template>
  <div class="cv-number-input" :class="`${carbonPrefix}--form-item`">
    <label :class="`${carbonPrefix}--label ${carbonPrefix}--skeleton`" />
    <div :class="`${carbonPrefix}--number ${carbonPrefix}--skeleton`" />
  </div>
</template>

<script setup>
import { carbonPrefix } from '../../global/settings';
</script>
