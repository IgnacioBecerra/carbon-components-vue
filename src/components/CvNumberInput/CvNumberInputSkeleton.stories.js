import {
  sbCompPrefix,
  storyParametersObject,
} from '../../global/storybook-utils';
import { StorybookGroupConst } from './StorybookGroupConst';
import CvNumberInputSkeleton from './CvNumberInputSkeleton.vue';

export default {
  title: `${sbCompPrefix}/${StorybookGroupConst}/CvNumberInputSkeleton`,
  component: CvNumberInputSkeleton,
};

const template = `<cv-number-input-skeleton></cv-number-input-skeleton>`;
const Template = (args, { argTypes }) => {
  return {
    props: Object.keys(argTypes),
    components: { CvNumberInputSkeleton },
    template,
    setup() {
      return { args };
    },
  };
};

export const Default = Template.bind({});
Default.parameters = storyParametersObject(
  Default.parameters,
  template,
  Default.args
);
