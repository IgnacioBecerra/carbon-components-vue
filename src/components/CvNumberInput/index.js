import CvNumberInput from './CvNumberInput.vue';
import CvNumberInputSkeleton from './CvNumberInputSkeleton.vue';

export { CvNumberInput, CvNumberInputSkeleton };
export default CvNumberInput;
