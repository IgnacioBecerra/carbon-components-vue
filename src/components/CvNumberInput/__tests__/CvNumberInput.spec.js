import { shallowMount, mount } from '@vue/test-utils';
import CvNumberInput from '../CvNumberInput.vue';

describe('CvNumberInput', () => {
  // ***************
  // SNAPSHOT TESTS
  // ***************

  it('should match snapshot when all props are default', async () => {
    const wrapper = await shallowMount(CvNumberInput);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match snapshot when it is not a formItem', async () => {
    const formItem = false;
    const wrapper = await shallowMount(CvNumberInput, { props: { formItem } });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match snapshot when all props are provided', async () => {
    const formItem = false;
    const helperText = 'helperText';
    const invalidMessage = 'invalidMessage';
    const label = 'label';
    const value = 5;
    const ariaLabelForDownButton = 'ariaLabelForDownButton';
    const ariaLabelForUpButton = 'ariaLabelForUpButton';
    const wrapper = await shallowMount(CvNumberInput, {
      props: {
        formItem,
        helperText,
        invalidMessage,
        label,
        value,
        ariaLabelForDownButton,
        ariaLabelForUpButton,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match snapshot when value is String', async () => {
    const formItem = false;
    const value = '15';
    const wrapper = await shallowMount(CvNumberInput, {
      props: { formItem, value },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match snapshot when value is Number', async () => {
    const formItem = false;
    const value = 5;
    const wrapper = await shallowMount(CvNumberInput, {
      props: { formItem, value },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match snapshot with min, max and step as Strings', async () => {
    const formItem = false;
    const value = '15';
    const min = '-10';
    const max = '10';
    const step = '2';
    const wrapper = await shallowMount(CvNumberInput, {
      props: { formItem, value, min, max, step },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match snapshot with optional min, max and step as numbers', async () => {
    const formItem = false;
    const value = 15;
    const min = -10;
    const max = 10;
    const step = 2;
    const wrapper = await shallowMount(CvNumberInput, {
      props: { formItem, value, min, max, step },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match snapshot when light theme', async () => {
    const formItem = false;
    const theme = 'light';
    const wrapper = await shallowMount(CvNumberInput, {
      props: { formItem, theme },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match snapshot when invalid-message slot is provided', async () => {
    const value = 5;
    const props = { value };
    const wrapper = await shallowMount(CvNumberInput, {
      slots: {
        'invalid-message': '<div class="fake-slot">my invalid message</div>',
      },
      props,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match snapshot when helper-text slot is provided', async () => {
    const value = 5;
    const props = { value };
    const wrapper = await shallowMount(CvNumberInput, {
      slots: {
        'helper-text': '<div class="fake-slot">my helper text</div>',
      },
      props,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
  const EmittedModelValueEvent = 'update:modelValue';
  const DefaultModelValue = 2;

  const testEmittedModelValue = async (wrapper, expectedValue) => {
    const modelValueEvent = wrapper.emitted(EmittedModelValueEvent);
    await expect(modelValueEvent).toBeTruthy();
    await expect(modelValueEvent[0][0]).toEqual(expectedValue);
  };

  const mountDefaultSetup = (modelValue = DefaultModelValue) => {
    return mount(CvNumberInput, { props: { modelValue } });
  };

  it('should emit input on doDown', async () => {
    const expectedValue = 1;

    const wrapper = mountDefaultSetup();
    const downButtonNode = wrapper.find('.down-icon');
    await downButtonNode.trigger('click');

    await testEmittedModelValue(wrapper, expectedValue);
  });

  it('should emit input on doUp', async () => {
    const expectedValue = 3;

    const wrapper = mountDefaultSetup();
    const upButtonNode = wrapper.find('.up-icon');
    await upButtonNode.trigger('click');

    await testEmittedModelValue(wrapper, expectedValue);
  });

  it('should emit input on change', async () => {
    const inputValue = 5;
    const expectedValue = inputValue;

    const wrapper = mountDefaultSetup();
    wrapper.find('input').setValue(inputValue);
    await wrapper.find('input').trigger('input');

    await testEmittedModelValue(wrapper, expectedValue);
  });

  it('should emit String when value prop is String', async () => {
    const modelValue = '555';
    const expectedValue = '556';

    const wrapper = mount(CvNumberInput, { props: { modelValue } });
    const upButtonNode = wrapper.find('.up-icon');
    await upButtonNode.trigger('click');

    await testEmittedModelValue(wrapper, expectedValue);
  });

  it('should emit value increased by 1 on doUp', async () => {
    const modelValue = 555;
    const expectedValue = 556;

    const wrapper = mount(CvNumberInput, { props: { modelValue } });
    const upButtonNode = wrapper.find('.up-icon');
    await upButtonNode.trigger('click');

    await testEmittedModelValue(wrapper, expectedValue);
  });

  it('should emit value decreased by 1 on doDown', async () => {
    const modelValue = 555;
    const expectedValue = 554;

    const wrapper = mount(CvNumberInput, { props: { modelValue } });
    const upButtonNode = wrapper.find('.down-icon');
    await upButtonNode.trigger('click');

    await testEmittedModelValue(wrapper, expectedValue);
  });

  it('should respect max value', async () => {
    const modelValue = 9;
    const step = 2;
    const max = 10;
    const expectedValue = max;

    const wrapper = mount(CvNumberInput, { props: { step, modelValue, max } });
    await wrapper.find('.up-icon').trigger('click');
    await wrapper.find('.up-icon').trigger('click');
    const modelValueEvent = wrapper.emitted(EmittedModelValueEvent);

    await expect(modelValueEvent[0][0]).toEqual(expectedValue);
    await expect(modelValueEvent[1][0]).toEqual(expectedValue);
  });

  it('should respect min value', async () => {
    const modelValue = 2;
    const step = 2;
    const min = 1;
    const expectedValue = min;

    const wrapper = await mount(CvNumberInput, {
      props: { step, modelValue, min },
    });
    await wrapper.find('.down-icon').trigger('click');
    await wrapper.find('.down-icon').trigger('click');
    const modelValueEvent = wrapper.emitted(EmittedModelValueEvent);

    await expect(modelValueEvent[0][0]).toEqual(expectedValue);
    await expect(modelValueEvent[1][0]).toEqual(expectedValue);
  });

  it('should emit the correct value on input', async () => {
    const modelValue = 2;
    const inputValue = 100;
    const expectedValue = inputValue;

    const wrapper = await mount(CvNumberInput, { props: { modelValue } });
    wrapper.find('input').setValue(inputValue);
    await wrapper.find('input').trigger('input');
    testEmittedModelValue(wrapper, expectedValue);
  });

  it('should watch modelValue', async () => {
    const modelValue = 2;
    const inputValue = 80;
    const expectedValue = inputValue;

    const wrapper = await mount(CvNumberInput, { props: { modelValue } });
    await wrapper.setProps({ modelValue: inputValue });

    expect(wrapper.vm.modelValue).toEqual(expectedValue);
  });
});
