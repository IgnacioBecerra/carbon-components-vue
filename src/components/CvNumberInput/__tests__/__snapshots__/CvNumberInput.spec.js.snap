// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvNumberInput should match snapshot when all props are default 1`] = `<cv-wrapper-stub tagtype="div" class="cv-number-input bx--form-item"></cv-wrapper-stub>`;

exports[`CvNumberInput should match snapshot when all props are provided 1`] = `<cv-wrapper-stub tagtype="" class="cv-number-input bx--form-item"></cv-wrapper-stub>`;

exports[`CvNumberInput should match snapshot when helper-text slot is provided 1`] = `<cv-wrapper-stub tagtype="div" class="cv-number-input bx--form-item"></cv-wrapper-stub>`;

exports[`CvNumberInput should match snapshot when invalid-message slot is provided 1`] = `<cv-wrapper-stub tagtype="div" class="cv-number-input bx--form-item"></cv-wrapper-stub>`;

exports[`CvNumberInput should match snapshot when it is not a formItem 1`] = `<cv-wrapper-stub tagtype="" class="cv-number-input bx--form-item"></cv-wrapper-stub>`;

exports[`CvNumberInput should match snapshot when light theme 1`] = `<cv-wrapper-stub tagtype="" class="cv-number-input bx--form-item"></cv-wrapper-stub>`;

exports[`CvNumberInput should match snapshot when value is Number 1`] = `<cv-wrapper-stub tagtype="" class="cv-number-input bx--form-item"></cv-wrapper-stub>`;

exports[`CvNumberInput should match snapshot when value is String 1`] = `<cv-wrapper-stub tagtype="" class="cv-number-input bx--form-item"></cv-wrapper-stub>`;

exports[`CvNumberInput should match snapshot with min, max and step as Strings 1`] = `<cv-wrapper-stub tagtype="" class="cv-number-input bx--form-item"></cv-wrapper-stub>`;

exports[`CvNumberInput should match snapshot with optional min, max and step as numbers 1`] = `<cv-wrapper-stub tagtype="" class="cv-number-input bx--form-item"></cv-wrapper-stub>`;
